src/wasm-binary.h

```diff
@@ -76,24 +79,37 @@ struct LEB {
     } while (more);
   }
 
-  void read(std::function<MiniT ()> get) {
+  void read(std::function<MiniT()> get) {
     value = 0;
     T shift = 0;
     MiniT byte;
     while (1) {
       byte = get();
-      value |= ((T(byte & 127)) << shift);
-      if (!(byte & 128)) break;
+      bool last = !(byte & 128);
+      T payload = byte & 127;
+      typedef typename std::make_unsigned<T>::type mask_type;
+      auto shift_mask = 0 == shift
+                            ? ~mask_type(0)
+                            : ((mask_type(1) << (sizeof(T) * 8 - shift)) - 1u);
+      T significant_payload = payload & shift_mask;
+      if (significant_payload != payload) {
+        assert(std::is_signed<T>::value && last &&
```

Comment:
Specifically, it is required to be a sign-extension of the value's sign bit, up to but not including the high bit of the final byte. So for a 32-bit value, there are 7 * 5 == 35 bits of data possible, but only 32 are valid. So the most significant byte of an encoded 5-byte LEB128 must be either `0b00000xxx` or `0b01111xxx`. If the LEB128 value is encoded in fewer than 5 bytes, then the sign bit is always `0b0100000`, so no sign-extension is necessary.